### PR TITLE
Updates the binary encoding and parameter cardinality sections

### PR DESCRIPTION
**Pull request by zslayton**
```
cargo install mdbook
cd ion-docs
git switch revamped-binary-enc
cd _books/ion-1-1
mdbook serve
```

----
_**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**_

### CODE REVIEW

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md:38]: Just a FYI or something to consider. In markdown, if you have two lines of text with no blank lines between them, a markdown renderer will consider them part of the same paragraph. In light of that, it is sometimes helpful to start new sentences on a new line to produce a cleaner diff. I sometimes try to do that myself, but it's not something I would ever block a PR for.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
When a macro parameter does not specify an encoding (i.e. the parameter name is not annotated), arguments
```
Or perhaps just "(the parameter name is not annotated)".

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]: nit: there are some opcodes that don't specify a length.
```suggestion
that dictates how to interpret the bytes that follow.
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
When the macro has a single _exactly-one_ parameter, the corresponding encoded argument follows the opcode and (if separate) the encoded address.
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
bytes in the stream. When no more parameters remain, parsing of the e-expression's arguments is complete.
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]: I think the numbers above the line of bytes are distracting from the main point of the diagram.
```suggestion
┌──── Opcode 0x00 is less than 0x40; this is an e-expression
│     invoking the macro at address 0.
│    ┌─── Argument 'a': opcode 0x61 indicates a 1-byte integer (1)
│    │     ┌─── Argument 'b': opcode 0x61 indicates a 1-byte integer (2)
│    │     │     ┌─── Argument 'c': opcode 0x61 indicates a 1-byte integer (3)
│  ┌─┴─┐ ┌─┴─┐ ┌─┴─┐
00 61 01 61 02 61 03
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
│    ┌─── Argument 'x': opcode 0x61 indicates a 1-byte integer (1)
│  ┌─┴─┐
00 61 01
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
│  │    ┌─── Argument 'x': opcode 0x61 indicates a 1-byte integer (1)
│  │  ┌─┴─┐
F4 01 61 01
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
│  │    ┌─── Argument 'a': opcode 0x61 indicates a 1-byte integer (1)
│  │    │     ┌─── Argument 'b': opcode 0x61 indicates a 1-byte integer (2)
│  │    │     │     ┌─── Argument 'c': opcode 0x61 indicates a 1-byte integer (3)
│  │  ┌─┴─┐ ┌─┴─┐ ┌─┴─┐
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md:267]: These could link to `values/float.md` instead. Your call, though.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]: minor style suggestion—this is slightly more technical and precise, and if you translate the specification to a language that reads in a different direction, it would still be clear.
```suggestion
The reader iterates over the parameters in the macro signature in the order they are declared. For each parameter,
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]: Did we settle on whether we're going to use `(% ...)` for variable expansion (#343)?
```suggestion
    (circle_area (radius) (multiply (.pi) (%radius) (%radius)))
```
That being said, this example might be confusing to newcomers. It looks like there could be a `.` missing in front of `multiply`. I can't think of another example right now, but if I can come up with something, perhaps you would consider changing it.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]: I think the `0b` is distracting, especially since one of the parameters is `b`.
```suggestion
| `(a b c?)`            | `------cc`                |
| `(a b* c?)`           | `----ccbb`                |
| `(a+ b* c?)`          | `--ccbbaa`                |
| `(a+ b c?)`           | `----ccaa`                |
| `(a+ b* c? d*)`       | `ddccbbaa`                |
| `(a+ b* c? d* e)`     | `ddccbbaa`                |
| `(a+ b* c? d* e f?)`  | `ddccbbaa ------ff`     |
| `(a+ b* c? d* e+ f?)` | `ddccbbaa ----ffee`     |
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]: I thought that `00` meant that the argument was an empty stream (not necessarily an empty expression group), and that `10` means that there is an expression group, but says nothing about how many expressions it contains (it could contain 0, hypothetically). Is that correct?

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]: Does "expression sequence" have a different meaning than "expression group"?

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]: I think having the numbers directly above the bytes is distracting from the main point here.
```suggestion
│  │    ┌── Argument 'a': opcode 0x61 indicates a 1-byte int (1)
│  │  ┌─┴─┐
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
│  │    ┌── Argument 'a': opcode 0x61 indicates a 1-byte int (1)
│  │  ┌─┴─┐
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
│  │  │    ┌── Opcode 0x61 indicates a 1-byte int (1)
│  │  │    │     ┌── Opcode 0x61 indicates a 1-byte int (2)
│  │  │    │     │     ┌─ Opcode 0x61 indicates a 1-byte int (3)
│  │  │  ┌─┴─┐ ┌─┴─┐ ┌─┴─┐
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (popematt) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
│  │  │    ┌── Opcode 0x61 indicates a 1-byte int (1)
│  │  │    │     ┌── Opcode 0x61 indicates a 1-byte int (2)
│  │  │    │     │     ┌─ Opcode 0x61 indicates a 1-byte int (3)
│  │  │    │     │     │    ┌─── Opcode 0xF0 is delimited end
│  │  │  ┌─┴─┐ ┌─┴─┐ ┌─┴─┐  │
00 02 01 61 01 61 02 61 03 F0
```

File: _books/ion-1-1/src/macros/defining_macros.md
Comment (popematt) [inline _books/ion-1-1/src/macros/defining_macros.md]: Is this supposed to be here?

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md:38]: When writing markdown, I find there's a tension between keeping the file tidy and keeping the diff tidy. :/ I once read a suggestion that every sentence in the file should be on its own line for optimal diffing; I really liked that, but found the result to occasionally be visually jarring.

It'd be cool if I could make an optimal-diff commit and then run an autoformatter as a final step to re-wrap the paragraphs to fit some number of columns.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md]: I find `i.e.` is less approachable than "that is", but YMMV. (I say that having already written "i.e." in a bunch of places. 😓)

> Or perhaps just "(the parameter name is not annotated)".

I like this version best, thanks.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md:267]: The encodings in `values/float.md` all start with the appropriate opcode; we'd have to add a section for the primitive encodings.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md]: Sure, I'll use that.

> and if you translate the specification to a language that reads in a different direction, it would still be clear

I'm not sure I follow. If the specification were in, say, Arabic, that wouldn't affect the direction that one would read the parameter names in the macro signature. They would still be left-to-right.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md]: Yeah, it's hard not to read it as math. The `.` was omitted intentionally--this example emits a somewhat arbitrary s-expression. I'll think of something else.

Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md]: I'm going to hold off on using `(%radius)` syntax because my next PR (in progress) updates variable expansion syntax throughout the spec.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md]: > I thought that 00 meant that the argument was an empty stream (not necessarily an empty expression group), 

You've pinpointed a distinction I was wrestling with while writing this! "Empty stream" is a much better term for this; as written, I had to coin a couple of other phrases to distinguish between populated/empty, implicit/explicit argument groups.

Let me take another pass at this using updated terminology.

File: _books/ion-1-1/src/macros/defining_macros.md
Comment (popematt) [inline _books/ion-1-1/src/macros/defining_macros.md]: I would not include this extension to the table. I understand what you're trying to get at, but I think it's best to keep the abstract model of macro arguments distinct from the encoding concerns for macro arguments.

My mental model of "Cardinality" is that it applies to the variable expansion. I.e. this is the essence of cardinality:
>If an argument expression expands to a number of values that the cardinality forbids, the reader must raise an error.


Then, restrictions on the encoding can be derived from the cardinality, and any such restrictions should be specified in the encoding sections (which you have done for example on L434 of `e_expressions.md`).

What do you think? Am I understanding things correctly?

Comment (zslayton) [reply _books/ion-1-1/src/macros/defining_macros.md]: > I think it's best to keep the abstract model of macro arguments distinct from the encoding concerns for macro arguments. My mental model of "Cardinality" is that it applies to the variable expansion.

Unfortunately, cardinality is a confluence of encoding and expansion. I wish that it weren't--my preference was that it only affect encoding (`!` or `*`) and that we'd denote expansion constraints separately (` -> string+`).  As it stands, the text encoding prohibits passing a populated expression group to `?` expressions. It also prohibits passing an empty expression group to ` +`. (Groups also cannot be passed to `!`, but that's not controversial.)

We need the text and binary encodings to be consistent enough in this regard that we can easily transcribe between the two. Either binary needs to prohibit the same combinations, or we allow variadics to use any encoding and only validate the expansion.

I'm open to removing this table extension for readability. However I think we should be _able_ to discuss the limitations at this spec level because they're consistent across the encodings. Ideally, reproducing the binary section's table in the text section would be redundant.

Let's talk about this more later today.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (tgregg) [inline _books/ion-1-1/src/binary/e_expressions.md]: I think we should either make it clear that `...` is not actual syntax, or we should use real syntax for the template body.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (tgregg) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
the bytes of the serialized arguments begin immediately after the e-expression opcode and (if separate) the macro address.
```
Or something like that, to be clear about which opcode we're referring to (since the arguments themselves don't have opcodes in this case).

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (tgregg) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
  (foo (flex_uint::a int8::b uint16::c) ...)
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (tgregg) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
limitation does not sacrifice any expressiveness; the desired constant can always be invoked directly
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (tgregg) [inline _books/ion-1-1/src/binary/e_expressions.md]:
```suggestion
construct: the _Argument Encoding Bitmap (AEB)_. This little-endian byte sequence precedes the first serialized argument
```

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (tgregg) [inline _books/ion-1-1/src/binary/e_expressions.md]: I suggest rewording this slightly to emphasize that the first byte(s) that follows the expression group's starting FlexUInt 0 byte is a `FlexUInt` length prefix. I think talking about the expressions first risks confusion.

File: _books/ion-1-1/src/macros/defining_macros.md
Comment (tgregg) [inline _books/ion-1-1/src/macros/defining_macros.md]: We could add a line to demonstrate how to use an empty expression group to elide x.

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md]: Yeah, that's fair. I'll replace each `...` with `/*...*/`--does that work?

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (zslayton) [reply _books/ion-1-1/src/binary/e_expressions.md]: Good catch

File: _books/ion-1-1/src/macros/defining_macros.md
Comment (zslayton) [reply _books/ion-1-1/src/macros/defining_macros.md]: I've [removed the extension](https://github.com/amazon-ion/ion-docs/pull/348/commits/45084a9685848198c5205a0878dc35fa2078deda) in the interest of getting the rest of this merged. We can revisit the table later.

**Review by popematt — APPROVED** [review]

File: _books/ion-1-1/src/binary/e_expressions.md
Comment (tgregg) [reply _books/ion-1-1/src/binary/e_expressions.md]: Yep!

**Review by tgregg — APPROVED** [review]

**Pull request merged**
